File: ruby/test/unit/parser_test.rb
require 'test_helper'

class ParserTest < Test::Unit::TestCase

  def setup
    @identifier = 'Item 1'
    @estimated_time = '5 hours'
    @priority = 'Urgent'
    @time = Time.parse('2013-09-03 15:09:22 +0200')
    @task = Todo::Task.new(@identifier, @estimated_time, @priority)
    @subject = Todo::Parser.new
  end

  def test_can_parse_open_task
    task = @subject.parse("* #{@identifier} -!- #{@estimated_time} -!- #{@priority}").first
    assert_equal @task.identifier, task.identifier
    assert_equal @task.estimated_time, task.estimated_time
    assert_equal @task.priority, task.priority
  end

  def test_can_parse_open_task_with_only_identifier
    task = @subject.parse("* #{@identifier} -!- -!- ").first
    assert_equal @task.identifier, task.identifier
    task = @subject.parse("* #{@identifier}").first
    assert_equal @task.identifier, task.identifier
  end

  def test_can_parse_comment
    assert_equal 1, @subject.parse('# this is a comment').length
    assert_equal 'this is a comment', @subject.parse('# this is a comment').first.text
    assert_equal "\# this is a comment\n", @subject.parse('# this is a comment').first.to_s
  end

  def test_can_parse_random_text
    assert_equal 4, @subject.parse("\n \# Comment \n\n this is just text").length
  end

  def test_can_parse_completed_task
    task = @subject.parse("- #{@identifier} -!- #{@estimated_time} -!- #{@priority} -!- #{@time}").first
    assert_equal @task.identifier, task.identifier
    assert_equal @task.estimated_time, task.estimated_time
    assert_equal @task.priority, task.priority
    assert_equal true, task.completed?
    assert_equal @time, task.completed_at
  end

  def test_can_parse_multiple_tasks
    assert_equal 5, @subject.parse("\n \# Comment \n\n * Item 1 -!- -!- \n - Item 2 -!- 3 minutes -!- Low -!- #{@time}").length
  end
end

File: java/src/test/java/me/ramblingsby/todos/ParserTest.java
package me.ramblingsby.todos;

import junit.framework.TestCase;

import java.util.List;

public class ParserTest extends TestCase {
  private String identifier, estimatedTime, priority, completedAt;
  private Task task;
  private Parser subject;

  public ParserTest(String testName) {
    super(testName);
    identifier = "Item 1";
    estimatedTime = "5 hours";
    priority = "Urgent";
    completedAt = "2013-09-03 15:09:22 +0200";
    task = new Task(identifier, estimatedTime, priority);
    subject = new Parser();
  }

  public void testCanParseOpenTask() throws Exception {
    Task parsedTask = (Task) subject.parse("* " + identifier + " -!- " + estimatedTime + " -!- " + priority).get(0);
    assertEquals(task, parsedTask);
  }

  public void testCanParseOpenTaskWithOnlyIdentifier() throws Exception {
    Task parsedTask = (Task) subject.parse("* " + identifier + " -!-  -!- ").get(0);
    assertEquals(identifier, parsedTask.getIdentifier());
    parsedTask = (Task) subject.parse("* " + identifier).get(0);
    assertEquals(identifier, parsedTask.getIdentifier());
  }

  public void testCanParseAComment() throws Exception {
    Comment comment = (Comment) subject.parse("# this is a comment").get(0);
    assertEquals("# this is a comment\n", comment.toString());
  }

  public void testCanParseRandomText() throws Exception {
    assertEquals(4, subject.parse("\n # Comment \n\n this is just text").size());
  }

  public void testCanParseCompletedTask() throws Exception {
    Task parsedTask = (Task) subject.parse("- " + identifier + " -!- " + estimatedTime + " -!- " + priority + " -!- " + completedAt).get(0);
    assertEquals(identifier, parsedTask.getIdentifier());
    assertEquals(estimatedTime, parsedTask.getEstimatedTime());
    assertEquals(priority, parsedTask.getPriority());
    assertEquals(true, parsedTask.completed());
    assertEquals(completedAt, parsedTask.getCompletedAt());
  }

  public void testCanParseMultipleTasks() throws Exception {
    List<Item> items = subject.parse("\n # Comment \n\n * Item 1 -!- -!- \n - Item 2 -!- 3 minutes -!- Low -!- " + completedAt);
    assertEquals(5, items.size());
  }
}